arroyo/processing/strategies/run_task_with_multiprocessing.py

```diff
@@ -483,24 +488,29 @@ def __check_for_results_impl(self, timeout: Optional[float] = None) -> None:
                 self.__next_step.submit(message)
 
             except MessageRejected:
-                result.next_index_to_process = idx
-                next_step_has_applied_backpressure = True
-                break
+                # For the next invocation of __check_for_results, start at this message
+                result.valid_messages_transformed.reset_iterator(idx)
 
-        if result.next_index_to_process != len(input_batch):
-            if next_step_has_applied_backpressure:
                 self.__metrics.increment(
                     "arroyo.strategies.run_task_with_multiprocessing.batch.backpressure"
                 )
-            else:
-                self.__metrics.increment(
-                    "arroyo.strategies.run_task_with_multiprocessing.batch.output.overflow"
+
+                # This is a warning because backpressure on the multiprocessing
+                # strategy _might_ cause degraded CPU saturation.
+                logger.warning(
+                    "Received backpressure from next_step, splitting output batch (%0.2f%% complete)",
+                    idx / len(input_batch) * 100,
                 )
+                raise NextStepTimeoutError()
```

Comment:
Would it be beneficial to add details around how much of the batch was processed before having to split it in the exception itself?

Reply:
it is logged in the warning, but lyn argues that we should have no warning. at the same time this exception is caught in the outer loop, it is effectively only for control flow